pyluna-pathology/luna/pathology/common/ml.py

```diff
@@ -150,4 +182,246 @@ def predict(self, input_tiles: torch.tensor):
+            preds = self.network(inputs)
+
+            # evaluate loss functional
+            loss = self.criterion(preds, labels)
+
+            # backward pass
+            self.optimizer.zero_grad()
+            loss.backward()
+            self.optimizer.step()
+
+            # compute metrics
+            epoch_loss += loss.item()
+            metrics(preds, labels)
+ 
+        # aggregate metrics accros batches, push to CPU
+        train_metrics = metrics.compute()
+        train_metrics = {k: v.cpu().numpy() for k, v in train_metrics.items()}
+
+        epoch_loss /= num_batches
+
+        train_metrics['train_loss'] = epoch_loss
+
+        return train_metrics
+        
+    def validate_epoch(self, dataloader: DataLoader, metrics: MetricCollection) -> Dict:
+        """a simple PyTorch validation loop that defines the model validation procedure
+        used during training. 
+
+        Args:
+            dataloader (DataLoader): a PyTorch dataloader object for the validation dataset
+            metrics (MetricCollection):  a collection of torchmetrics used to evaluate model performance
+        Returns:
+            Dict: A dictionary of metrics used for monitoring and model evaluation 
+        """
+        print("starting validation")
+
+        num_batches = len(dataloader)
+        self.network.eval()
+        loss = 0 
+
+        with torch.no_grad():
+            for epoch_iter, (inputs, labels) in enumerate(dataloader):
+
+                # forward pass
+                preds = self.network(inputs)
+
+                # compute validation loss
+                loss += self.criterion(preds, labels).item()
+                
+                # compute metrics 
+                metrics(preds, labels)
+        
+        # aggregate metrics across batches, push to CPU
+        val_metrics = metrics.compute()
+        val_metrics = {k: v.cpu().numpy() for k, v in val_metrics.items()}
+
+        loss /= num_batches
+
+        val_metrics['val_loss'] = loss
+
+        return val_metrics
+
+
+class StratifiedGroupKFold(_BaseKFold):
+    """Stratified K-Folds iterator variant with non-overlapping groups.
+    See https://github.com/scikit-learn/scikit-learn/issues/13621 for a
+    full discussion on group stratified cross validation.
+
+    This cross-validation object is a variation of StratifiedKFold that returns
+    stratified folds with non-overlapping groups. The folds are made by
+    preserving the percentage of samples for each class.
+
+    The same group will not appear in two different folds (the number of
+    distinct groups has to be at least equal to the number of folds).
+
+    The difference between GroupKFold and StratifiedGroupKFold is that
+    the former attempts to create balanced folds such that the number of
+    distinct groups is approximately the same in each fold, whereas
+    StratifiedGroupKFold attempts to create folds which preserve the
+    percentage of samples for each class.
+    """
+
```

Comment:
Oh, huh. I don't remember StratifiedGroupKFold being in sklearn when I originally made this, I think it might've been a PR or in the issues at that time. I'll use the sklearn implementation then.